### PR TITLE
Ad/test/win32 smoke #1092

### LINKED ISSUE #1092: Smoke test for Win32
Nuget smoke test to verify that basics work with a simple Win32 app

### PR DESCRIPTION
**Pull request by AndyDentFree**
Simple standalone app just using Realm.Database. Anyone can review.
Resolves #1092

### CODE REVIEW

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                            new Employee {Name="Jake"}
+                        }
+                });
+                foreach (var emp in theBoss.Reports)
+                {
+                    emp.Boss = theBoss;
+                }
+            });
+            var jakes = from d in realm.All<Employee>() where d.Name == "Jake" select d;
+            var jake = realm.All<Employee>().Single(p => p.Name == "Jake") as Employee;
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs]: Why the cast?

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            Console.WriteLine("About to test Realm on desktop:");
+            var msg = TestRealm.SmokeTest();
+            Console.WriteLine(msg);
+            Console.WriteLine("Press any key to continue");
+            var ignored = Console.ReadKey();
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs]: Why assign to a variable?

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("SmolkeTestWin32Console")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("SmolkeTestWin32Console")]
+[assembly: AssemblyCopyright("Copyright ©  2017")]
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs]: These should be removed and we should add a link to `RealmAssemblyInfo`

File: wrappers/README.md [hunk 4]
```diff
@@ -30,8 +30,8 @@ If you downloaded a zip of the source, you need to go back to github to identify
 1. Download a zip using the GitHub download button in that tree, eg `realm-object-store-fb2ed6aa0073be4cb0cd059cae407744ee883b77.zip`
 1. Unpack its contents into `wrappers/src/object-store`
 
-Building Wrappers on macOS
---------------------------
+Building iOS and Android Wrappers on macOS
+------------------------------------------
```
Comment (nirinchev) [inline wrappers/README.md]: Why the extra dashes? Wouldn't it render the same?

File: Platform.Win32/Tests.Win32/Tests.Win32.csproj [hunk 5]
```diff
@@ -3,7 +3,7 @@
   <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
```
Comment (nirinchev) [inline Platform.Win32/Tests.Win32/Tests.Win32.csproj]: Why is this change needed?

File: Tests/SmokeTestWin32Console/README.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# About SmokeTestWin32Console
+
+This is a very simple program showing a Realm instance being loaded, data written and queried, in a pure Console app.
+
+It builds for x86 and x64.
+
+Note that it only uses the standalone Realm.Database nuget.
+
+The project was created and tested under Visual Studio 2017 RC and may run into minor project issues with earlier versions.
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/README.md]: What issues? I would expect a simple console app to work in at least VS 2015.

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs (hunk 1 above)
Comment (AndyDentFree) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs]: Habit from elsewhere, fixed it.

File: wrappers/README.md (hunk 4 above)
Comment (AndyDentFree) [inline wrappers/README.md]: Yes it renders the same but it is more pleasing to the eye in plain text mode.

File: Platform.Win32/Tests.Win32/Tests.Win32.csproj (hunk 5 above)
Comment (AndyDentFree) [inline Platform.Win32/Tests.Win32/Tests.Win32.csproj]: It is a lot clearer that people are selecting x86 or x64 builds because **they have to** - any time we can reduce the impression that we actually have a pure _AnyCPU_ product I think we should do so.

File: Tests/SmokeTestWin32Console/README.md (hunk 6 above)
Comment (AndyDentFree) [inline Tests/SmokeTestWin32Console/README.md]: I don't know but I had some weirdness when I tried going back to VS2015 - IDE crashed.

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (AndyDentFree) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs]: No, that introduces too much cross-dependency. Values fixed, thanks.

File: wrappers/README.md (hunk 4 above)
Comment (nirinchev) [inline wrappers/README.md]: Then I guess https://github.com/realm/realm-dotnet/pull/1164/files/80880d8ec0b4d69f733a6a486a7044f3e9913528#diff-bebcb6e63d811c0b5f18fba9b7a7da3eR50 should be longer

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("SmolkeTestWin32Console")]
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs]: Typo

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+[assembly: AssemblyDescription("Small test program for Realm")]
+[assembly: AssemblyCompany("Realm Inc.")]
+[assembly: AssemblyProduct("Realm C#")]
+[assembly: AssemblyCopyright("Copyright © 2017 Realm")]
+
+
+// Setting ComVisible to false makes the types in this assembly not visible
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Properties/AssemblyInfo.cs]: These comments are not needed

File: Platform.Win32/Tests.Win32/Tests.Win32.csproj (hunk 5 above)
Comment (fealebenpae) [inline Platform.Win32/Tests.Win32/Tests.Win32.csproj]: Our assembly is AnyCPU, though.

Comment (AndyDentFree) [inline Platform.Win32/Tests.Win32/Tests.Win32.csproj]: >Our assembly is AnyCPU, though.

Then that is a separate bug.

If you have an assembly with a dependency on native that **requires** it to be built as x86 or x64 it should never be built and shipped as _AnyCPU_

Comment (fealebenpae) [inline Platform.Win32/Tests.Win32/Tests.Win32.csproj]: Why? We provide both x86 and x64 builds of our native libraries and the correct one is loaded at runtime. Why does the managed assembly have to be architecture specific?

**Comment by AndyDentFree** [comment]
retest this please

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                            new Employee {Name="Sree" },
+                            new Employee {Name="Jake"}
+                        }
+                });
+                foreach (var emp in theBoss.Reports)
+                {
+                    emp.Boss = theBoss;
+                }
+            });
+            var jakes = from d in realm.All<Employee>() where d.Name == "Jake" select d;
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs]: This doesn't seem to be used.

File: Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+namespace SmokeTestWin32Console
+{
+
+    public class Employee : RealmObject
+    {
+        public string Name { get; set; }
+
+        public IList<Employee> Reports { get; }
+
+        // circular reference breaks schema see issue 1163 [Backlink(nameof(Employee.Reports))]
```
Comment (nirinchev) [inline Tests/SmokeTestWin32Console/SmokeTestWin32Console/Program.cs]: Should be fixed now.

**Comment by AndyDentFree** [comment]
Closing, replaced by a copy which managed to get through CI without weird crashes in unrelated areas.
